Lib/typing.py

```diff
@@ -308,19 +308,67 @@ def _unpack_args(args):
             newargs.append(arg)
     return newargs
 
-def _deduplicate(params):
+def _deduplicate(params, *, unhashable_fallback=True):
     # Weed out strict duplicates, preserving the first of each occurrence.
-    all_params = set(params)
-    if len(all_params) < len(params):
-        new_params = []
-        for t in params:
-            if t in all_params:
-                new_params.append(t)
-                all_params.remove(t)
-        params = new_params
-        assert not all_params, all_params
-    return params
-
+    try:
+        all_params = set(params)
+    except TypeError:
+        if not unhashable_fallback:
+            raise
+
+        # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
+        hasable, unhashable = _deduplicate_unhashable(params)
+        return hasable + unhashable
+
+    if len(all_params) == len(params):  # fast case
+        return params
+    new_params = []
+    for t in params:
+        if t in all_params:
+            new_params.append(t)
+            all_params.remove(t)
+    assert not all_params, all_params
+    return new_params
+
+def _deduplicate_unhashable(params):
+    regular_params = []
+    unhashable_params = []
+    for param in params:
+        try:
+            hash(param)
+        except TypeError:
+            unhashable_params.append(param)
+        else:
+            regular_params.append(param)
+
+    unique_params = set(regular_params)
+    new_hashable = []
+    for t in regular_params:
+        if t in unique_params:
+            new_hashable.append(t)
+            unique_params.remove(t)
+    assert not unique_params, unique_params
+
+    new_unhashable = []
+    for t in unhashable_params:
+        # This is slow, but this is the only safe way to compare
+        # two `Annotated[]` instances, since `metadata` might be unhashable.
```

Comment:
This comment seems inapposite, as this path isn't only for Annotated, and your tests don't even use Annotated.